src/main/java/org/assertj/core/internal/Maps.java

```diff
@@ -370,15 +363,67 @@ public void assertHasSameSizeAs(AssertionInfo info, Map<?, ?> map, Map<?, ?> oth
     }
     failIfEmpty(keys, keysToLookForIsEmpty(placeholderForErrorMessages));
 
-    Set<K> notFound = new LinkedHashSet<>();
-    Set<K> notExpected = new LinkedHashSet<>();
-
-    compareActualMapAndExpectedKeys(actual, keys, notExpected, notFound);
+    Set<K> notFound = getNotFoundKeys(actual, keys);
+    Set<K> notExpected = getNotExpectedKeys(actual, keys);
 
     if (!notFound.isEmpty() || !notExpected.isEmpty())
       throw failures.failure(info, shouldContainOnlyKeys(actual, keys, notFound, notExpected));
   }
 
+  private static <K> Set<K> getFoundKeys(Map<K, ?> actual, K[] expectedKeys) {
+    Set<K> found = new LinkedHashSet<>();
+    for (K expectedKey : expectedKeys) {
+      if (actual.containsKey(expectedKey)) found.add(expectedKey);
+    }
+    return found;
```

Comment:
could use Stream API filter:
```java
return Stream.of(expectedKeys)
             .filter(actual::containsKey) 
             .collect(toCollection( LinkedHashSet::new));
```

Reply:
Initially, I used the Stream API but @filiphr correctly pointed out that the for loop would be better performance-wise. I seem to recall some issues about performance in case of big input, so the proposal seemed valid to me. What do you think about it?

Reply:
fair enough, let's add a comment saying we don't use stream for performance reasons.